Add "Auto" value in option parsing

wthrowe opened the pull request:
## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Add Auto class for option parsing

wthrowe committed:
Use Options::Auto for CCE time range

nilsvu commented on src/Options/Auto.hpp:
```diff
@@ -0,0 +1,64 @@
+
+namespace Options {
+/// \ingroup OptionParsingGroup
+/// A class indicating that a parsed value can be automatically
```
This paragraph should probably be `\brief`

nilsvu commented on src/Options/Auto.hpp:
```diff
@@ -0,0 +1,64 @@
+template <typename T>
+class Auto {
+ public:
+  Auto() = default;
```
noexcept?

wthrowe replied:
The meaning of writing noexcept there varies between versions of the standard, and clang implements it incorrectly for C++17, so I'm avoiding it these days.  Leaving it out always does the right thing.  (Same with constexpr.)

nilsvu commented on src/Options/Auto.hpp:
```diff
@@ -0,0 +1,64 @@
+
+#include <optional>
+#include <ostream>
+#include <string>
```
include `utility`

nilsvu commented on src/Options/Auto.hpp:
```diff
@@ -0,0 +1,64 @@
+};
+
+template <typename T>
+std::ostream& operator<<(std::ostream& s, const Auto<T>& x) noexcept {
```
(nitpick) in other classes we call `s` -> `os`

nilsvu commented on tests/Unit/Options/Test_Auto.cpp:
```diff
@@ -0,0 +1,156 @@
+  check_create<std::vector<int>>(
+      "- 1\n"
+      "- 2\n"
+      "- 3\n",
```
you could also write this as `"[1, 2, 3]"`

nilsvu reviewed (changes requested):
This looks great! A few minor comments:

nilsvu commented:
LGTM please squash

wthrowe commented:
Squashed.  I also updated the second commit to also do the CCE end time, which I missed before.  The changes were virtually identical to those for the start time (and not very long), so I squashed them in.

moxcodes reviewed (dismissed):
Looks great, thanks for adding this!

nilsdeppe commented:
The build fails (I think partly because the compilers are being silly), and there is an actual clang-tidy warning

wthrowe commented:
The fixes for CI were at least slightly nontrivial, so I pushed them as a fixup.

nilsdeppe commented:
Looks good. I'm happy for you to squash :)

moxcodes reviewed (dismissed):
I'm happy for you to squash.

nilsvu commented on tests/Unit/Options/Test_Auto.cpp:
```diff
@@ -0,0 +1,153 @@
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif  // defined(__GNUC__) && !defined(__clang__) && __GNUC__ => 8 && __GNUC__ < 11
+    static type default_value() noexcept { return {}; }
```
- Do you have an idea why GCC doesn't like this instantiation, but is fine with the one in the CCE code? It's a bit unfortunate that the pragmas get printed to the documentation snippet.
- You add pragmas for GCC 8--9 in Auto.hpp, and for GCC 8--10 here. Is that on purpose?

Please squash directly if you end up making any changes.

wthrowe replied:
I have no idea why different version of GCC complain about different variations on that line.

Good point about the documentation.  I've moved the pragmas outside the snippet.

nilsvu replied:
So GCC 10 indeed complains about the one line but not the other, whereas GCC 8 & 9 complain about both? that's weird
